service/s3/s3crypto/s3crypto_client.go

```diff
@@ -0,0 +1,217 @@
+package s3crypto
+
+// DR TOOLS TEST integration test bucket
+import (
+	"bytes"
+	"encoding/base64"
+	"io/ioutil"
+	"math/rand"
+
+	"github.com/aws/aws-sdk-go/aws/client"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+const (
+	ivSize  = 16
+	keySize = 32
+)
+
+// Client supports client level encryption and decryption to S3.
+// The crypto client uses "envelope" encryption".
+type Client struct {
+	s3client *s3.S3
+	cfg      Config
+}
+
+// Config used to customize the Client
+type Config struct {
+	// MasterCipher will hold either a master symmetric key or a KMS client
+	// for encrypting the envelope key
+	MasterCipher Cipher
+}
+
+// NewClient will return a crypto client and set an aes cbc encrypter
+// as one of the fields.
+func NewClient(p client.ConfigProvider, options ...func(*Config)) (*Client, error) {
+	client := &Client{}
+
+	client.s3client = s3.New(p)
+
+	for _, option := range options {
+		option(&client.cfg)
+	}
+
+	// TODO: If MasterCipher hasn't been set error out
+	return client, nil
+}
+
+// PutObjectInput for the PutObjectRequest
+type PutObjectInput struct {
+	SaveStrategy
+	MaterialDesc     string
+	S3PutObjectInput *s3.PutObjectInput
+}
+
+// PutObjectRequest will call the S3's PutObjectRequest and then save the envelope
+// to the appropriate place.
+func (client *Client) PutObjectRequest(input *PutObjectInput) (*request.Request, *s3.PutObjectOutput) {
+	req, out := client.s3client.PutObjectRequest(input.S3PutObjectInput)
+
+	iv := generateRandBytes(ivSize)
+	key := generateRandBytes(keySize)
+	// TODO: Let customers specify the cipher
+	enc, err := NewAESCBC(key, iv)
```

Comment:
Until the customer is allowed to specify the cipher, consider defaulting to AES-GCM rather than AES-CBC. AES-GCM (an AEAD) is generally considered safer.